### PR TITLE
graciously handle array shuffling

### LINKED ISSUE #4412: preact@10.22.0 unexpected child ordering with skew based diff
- [ ] Check if updating to the latest Preact version resolves the issue

**Describe the bug**
With preact version from 10.16.0 to 10.22.0, and code below:

```jsx
<div>
  {items.map((key) => (
    <div key={key}>{key}</div>
  ))}
</div>
```
Updating array `items` (shuffle it) leads to unexpected ordering.

preact version 10.15.0 and below will not reproduce this issue.

**To Reproduce (with Stackblitz)**

1. Open project at https://stackblitz.com/edit/create-preact-starter-ckjsze?file=src%2Findex.jsx
2. You can see numbers in order:
    ```
    1
    3
    4
    0
    5
    2
    6
    ```

**Expected behavior**
Numbers should be in order:
```
1
3
5
2
6
4
0
```

**To Reproduce (in preact unit test)**

1. Put code below inside file `test/browser/render.test.js`,
    ```jsx
    it("shuffle", function () {
	    const jsx = (items) => (
		    <div>
			    {items.map((key) => (
				    <div key={key}>{key}</div>
			    ))}
		    </div>
	    );
	    const a = ["0", "1", "2", "3", "4", "5", "6"];
	    const b = ["1", "3", "5", "2", "6", "4", "0"];
	    render(jsx(a), scratch);
	    render(jsx(b), scratch);
    
	    expect(scratch.innerHTML).to.equal(
		    `<div>${b.map(n => `<div>${n}</div>`).join("")}</div>`
	    );
    });
    ```
2. run test, the test above will FAIL

### PR DESCRIPTION
**Pull request by JoviDeCroock**
Fixes #4412 

I will have to do a second pass to make these comments better because every time I dive through this in search of bugs I have to re-ingest this whole algorithm

### CODE REVIEW

File: src/diff/children.js [hunk 1]
```diff
@@ -302,20 +302,34 @@ function constructNewChildrenArray(newParentVNode, renderResult, oldChildren) {
...
+				// than where it's at now.
+
+				// When the remaining old chiildren is less than the new children
+				// plus our skewed index we know we are dealing with a growing list
+				if (remainingOldChildren < newChildrenLength + skewedIndex) {
+					skew += matchingIndex + skewedIndex;
+				} else {
+					skew = 0;
 				}
-			} else {
```
Comment (JoviDeCroock) [inline src/diff/children.js:317]: This else branch was unreachable, we check whether we are in a 1 offset first and then lesser or larger

File: test/browser/render.test.js [hunk 2]
```diff
@@ -1625,4 +1625,31 @@ describe('render()', () => {
 			'<div><div>One</div><div>Six</div><div>Seven</div></div>'
 		);
 	});
+
+	it('shuffle', function () {
```
Comment (JoviDeCroock) [inline test/browser/render.test.js:1629]:
```suggestion
	it('handles shuffled child-ordering', function () {
```

File: test/browser/render.test.js [hunk 3]
```diff
@@ -474,17 +474,17 @@ describe('render()', () => {
...
 	it('should support popover true boolean', () => {
 		render(<div popover />, scratch);
-		expect(scratch.innerHTML).to.equal("<div popover=\"\"></div>");
+		expect(scratch.innerHTML).to.equal('<div popover=""></div>');
 	});
 
 	it('should support popover false boolean', () => {
 		render(<div popover={false} />, scratch);
-		expect(scratch.innerHTML).to.equal("<div></div>");
+		expect(scratch.innerHTML).to.equal('<div></div>');
```
Comment (JoviDeCroock) [inline test/browser/render.test.js:487]: Linter auto-correct

File: test/browser/render.test.js [hunk 4]
```diff
@@ -1625,4 +1625,31 @@ describe('render()', () => {
...
+		);
+		const a = ['0', '1', '2', '3', '4', '5', '6'];
+		const b = ['1', '3', '5', '2', '6', '4', '0'];
+		render(<App items={a} />, scratch);
+		expect(scratch.innerHTML).to.equal(
+			`<div>${a.map(n => `<div>${n}</div>`).join('')}</div>`
+		);
+
+		render(<App items={b} />, scratch);
+
```
Comment (JoviDeCroock) [inline test/browser/render.test.js:1645]:
```suggestion
```

File: src/diff/children.js [hunk 5]
```diff
@@ -302,20 +302,34 @@ function constructNewChildrenArray(newParentVNode, renderResult, oldChildren) {
...
-					skew = matchingIndex - skewedIndex;
+				// Our matched DOM-node is further in the negative way in the list of children
+				// than where it's at now.
+
+				// When the remaining old chiildren is less than the new children
+				// plus our skewed index we know we are dealing with a growing list
+				if (remainingOldChildren < newChildrenLength + skewedIndex) {
+					skew += matchingIndex + skewedIndex;
+				} else {
+					skew = 0;
```
Comment (JoviDeCroock) [inline src/diff/children.js:331]: I took this from the v11/keyed branch but I am not a 100% sure on this

**Review by marvinhagemeister — APPROVED** [review]

**Pull request merged**